### PR TITLE
Extract ion names from ccd

### PR DESCRIPTION
**Pull request by McMenemy**
Creates a script to extract Ion Names - [Issue 62](https://github.com/molstar/molstar/issues/62).

There are a few differences between the old hardcoded IonNames and the new extracted IonNames:
inOldNotNew = [ 'MLT', 'NCO', 'OHX' ]
inNewNotOld = [ '4TI', 'CF', 'ZR' ] 

Command to generate ions.ts ```$ node --max-old-space-size=4096 lib/commonjs/cli/chem-comp-dict/create-table.js build/data/ccb.bcif -ino src/mol-model/structure/model/types/ions.ts```. Let me know if you think it's cleaner to move this to its own separate file instead of extending ```create-table.js```.

Example File output
```
/**
* Copyright (c) 2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
*
* Code-generated ion names params file. Names extracted from CCD components.
*
* @author molstar/chem-comp-dict/create-table cli
*/

export const IonNames = new Set(['118', '119', '543', '1AL', '1CU', '2FK', '2HP', '2OF', '3CO', '3MT', '3NI', '3OF', '3P8', '4MO', '4PU', '4TI', '6MO', 'ACT', 'AG', 'AL', 'ALF', 'AM', 'ATH', 'AU', 'AU3', 'AUC', 'AZI', 'BA', 'BCT', 'BEF', 'BF4', 'BO4', 'BR', 'BS3', 'BSY', 'CA', 'CAC', 'CD', 'CD1', 'CD3', 'CD5', 'CE', 'CF', 'CHT', 'CL', 'CO', 'CO3', 'CO5', 'CON', 'CR', 'CS', 'CSB', 'CU', 'CU1', 'CU3', 'CUA', 'CUZ', 'CYN', 'DME', 'DMI', 'DSC', 'DTI', 'DY', 'E4N', 'EDR', 'EMC', 'ER3', 'EU', 'EU3', 'F', 'FE', 'FE2', 'FPO', 'GA', 'GD3', 'GEP', 'HAI', 'HG', 'HGC', 'IN', 'IOD', 'IR', 'IR3', 'IRI', 'IUM', 'K', 'KO4', 'LA', 'LCO', 'LCP', 'LI', 'LU', 'MAC', 'MG', 'MH2', 'MH3', 'MLI', 'MMC', 'MN', 'MN3', 'MN5', 'MN6', 'MO1', 'MO2', 'MO3', 'MO4', 'MO5', 'MO6', 'MOO', 'MOS', 'MOW', 'MW1', 'MW2', 'MW3', 'NA', 'NA2', 'NA5', 'NA6', 'NAO', 'NAW', 'NET', 'NH4', 'NI', 'NI1', 'NI2', 'NI3', 'NO2', 'NO3', 'NRU', 'O4M', 'OAA', 'OC1', 'OC2', 'OC3', 'OC4', 'OC5', 'OC6', 'OC7', 'OC8', 'OCL', 'OCM', 'OCN', 'OCO', 'OF1', 'OF2', 'OF3', 'OH', 'OS', 'OS4', 'OXL', 'PB', 'PBM', 'PD', 'PDV', 'PER', 'PI', 'PO3', 'PO4', 'PR', 'PT', 'PT4', 'PTN', 'RB', 'RH3', 'RHD', 'RU', 'SB', 'SCN', 'SE4', 'SEK', 'SM', 'SMO', 'SO3', 'SO4', 'SR', 'T1A', 'TB', 'TBA', 'TCN', 'TEA', 'TH', 'THE', 'TL', 'TMA', 'TRA', 'UNX', 'V', 'VN3', 'VO4', 'W', 'WO5', 'Y1', 'YB', 'YB2', 'YH', 'YT3', 'ZCM', 'ZN', 'ZN2', 'ZN3', 'ZNO', 'ZO3', 'ZR']);
```

### CODE REVIEW

File: src/cli/chem-comp-dict/create-table.ts [hunk 1]
```diff
@@ -243,8 +243,37 @@ function createAtoms(ccd: DatabaseCollection<CCD_Schema>) {
     );
 }
 
-async function run(out: string, binary = false, ccaOut?: string) {
-    const { bonds, atoms } = await createBonds(!!ccaOut);
+function extractIonNames(ccd: DatabaseCollection<CCD_Schema>) {
+    const ionNames: string[] = [];
+    for (const k in ccd) {
+        const {chem_comp} = ccd[k];
+        if (chem_comp.name.value(0).toUpperCase().includes(' ION')) {
```
Comment (dsehnal) [inline src/cli/chem-comp-dict/create-table.ts:250]: Are you sure this is a necessary and/or correct condition?

I know this wasn't included in the original list, but should this include anions/cations as well? See for example http://files.rcsb.org/ligands/view/FLC.cif

Also this is PDB, so I would not be surprised if there were names with "-ION" or other variants.

**Review by dsehnal — COMMENTED** [review]
Thanks for the contribution.

File: src/cli/chem-comp-dict/create-table.ts (hunk 1 above)
Comment (arose) [reply src/cli/chem-comp-dict/create-table.ts:250]: Hi David, he based it on an SQL query I used previously to get a list of small ions. Heavy atom count should be below 8 so something like https://www.rcsb.org/ligand/OHX would be the largest.

**Comment by arose** [comment]
Thanks @McMenemy and great for you to give a diff of the old and new list. I wonder how MLT made it into old list, should not be there. NCO and OHX should be, could you please add them as extra names in the script itself? The three new ones are good, they were added to the PDB archive after I created the list.

A seperate script in the same directory reusing code from the existing one would be great.

**Comment by McMenemy** [comment]
@arose I moved the functionality to it's own script with 'NCO' and 'OHX' hardcoded as ions. 

@dsehnal I was thinking more about your comment on figuring out a better way to determine ions. The best I came up with is to make sure that the _chem_comp.pdbx_formal_charge is not 0 and that the _chem_comp.formula has less than 8 heavy atoms (eg. "C5 H14 B N O3" == 10 heavy atoms). I can implement and test that solution if everyone thinks it would work better.

Thank you both for the quick feedback.

File: src/cli/chem-comp-dict/create-ions.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env node
+/**
+ * Copyright (c) 2018-2019 mol* contributors, licensed under MIT, See LICENSE file for more info.
```
Comment (arose) [inline src/cli/chem-comp-dict/create-ions.ts:3]: nitpick: 2020

**Comment by arose** [comment]
This is great, thanks again @McMenemy.

Can you please check that `_chem_comp.pdbx_release_status` is REL (for released). There are currently quite a few obsolete (OBS) entries in the list.

Requiring _chem_comp.pdbx_formal_charge to be non zero is too strict.

Using the `_chem_comp.formula` to check heavy atom count sounds (I was thinking to loop over the atom table but this seems shorter).

**Comment by McMenemy** [comment]
Adding the REL filter removed the following 53 ions: 

```
[
  '119', '1CU', '2OF', '3OF', '543',
  'CD1', 'CD3', 'CD5', 'CO5', 'E4N',
  'HGC', 'KO4', 'MH3', 'MN5', 'MN6',
  'MO1', 'MO2', 'MO3', 'MO4', 'MO5',
  'MO6', 'MW1', 'MW2', 'MW3', 'NA2',
  'NA5', 'NA6', 'NAO', 'NAW', 'NI1',
  'NI2', 'NI3', 'O4M', 'OC1', 'OC2',
  'OC3', 'OC4', 'OC5', 'OC6', 'OC7',
  'OC8', 'OCL', 'OCM', 'OCN', 'OCO',
  'OF1', 'OF2', 'OF3'
]
```

**Comment by dsehnal** [comment]
Is keeping the obsolete entries in the list actually an issue? What if someone wants to open say an older trajectory file?

**Comment by arose** [comment]
> Is keeping the obsolete entries in the list actually an issue? What if someone wants to open say an older trajectory file?

You are right, we should keep them. Should not be an issue, the names will not get reused for non ions or anything else.

**Comment by McMenemy** [comment]
Makes sense, I reverted the REL filter.

**Comment by arose** [comment]
Thank you very much @McMenemy!

When I tried it, the create-table script would also run, so I moved the shared code to a util file, see 17162e967ad3194ed7629179a10b8d25298160f9.

**Pull request merged**